File: p74.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    bool binary_search(vector<int> &vec, int target, int start, int end) {
        if(target < vec[start] || target > vec[end])
            return false;
        int mid = (end+start)/2;
        if(vec[mid] == target)
            return true;
        else if(vec[mid] > target)
            return binary_search(vec, target, start, mid-1);
        else
            return binary_search(vec, target, mid+1, end);
    }
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if(matrix.size() == 0 || matrix[0].size() == 0)
            return false;
        int m = matrix.size();
        int n = matrix[0].size();
        // find the row that contains our number
        for(int i = 0; i < m; ++i) {
            // we found the row
            if(matrix[i][0] <= target && matrix[i][n-1] >= target) {
                // binary search on the row
                return binary_search(matrix[i], target, 0, n-1);
            }
        }
        // it's out of range of all of our rows
        return false;
    }
};

File: p74.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        def bin_search(arr, n):
            if len(arr) == 0:
                return False

            left, right = 0, len(arr)
            mid = (left+right)//2

            while left < mid:
                if arr[mid] > n:
                    right = mid
                else:
                    left = mid
                mid = (left+right)//2

            return arr[left] == n

        left, right = 0, len(matrix)
        mid = (left+right)//2

        while left < mid:
            if matrix[mid][0] <= target <= matrix[mid][-1]:
                return bin_search(matrix[mid], target)
            elif matrix[mid][0] > target:
                right = mid
            else:
                left = mid+1
            mid = (left+right)//2

        return 0 <= mid < len(matrix) and bin_search(matrix[mid], target)
